compiler/sempass2.nim

```diff
@@ -259,6 +269,33 @@ proc listGcUnsafety(s: PSym; onlyWarning: bool; conf: ConfigRef) =
   var cycleCheck = initIntSet()
   listGcUnsafety(s, onlyWarning, cycleCheck, conf)
 
+proc listSideEffects(s: PSym; onlyWarning: bool; cycleCheck: var IntSet; conf: ConfigRef) =
+  for u in s.sideEffectReasons:
+    if u != nil and not cycleCheck.containsOrIncl(u.id):
+      let msgKind = if onlyWarning: warnSideEffect else: errGenerated
+      case u.kind
+      of skLet, skVar:
+        message(conf, s.info, msgKind,
+          ("'$#' accesses global state at '$#'") % [s.name.s, u.name.s])
+      of routineKinds:
+        # recursive call *always* produces only a warning so the full error
+        # message is printed:
+        listSideEffects(u, true, cycleCheck, conf)
+        message(conf, s.info, msgKind,
+          "'$#' has side effects as it calls '$#'" %
+          [s.name.s, u.name.s])
+      of skParam, skForVar:
+        message(conf, s.info, msgKind,
+          "'$#' has side effects as it performs an indirect call via '$#'" %
+          [s.name.s, u.name.s])
+      else:
+        message(conf, u.info, msgKind,
```

Comment:
generate a single message instead and get rid of `onlyWarning: bool; ` which simplifies and renders better:
```nim
proc listSideEffects(result: var string, s: PSym; cycleCheck: var IntSet; conf: ConfigRef) = ... # ok, still recursive

var msg = ""
listSideEffects(msg, ...)
message(conf, n.info, errGenerated, msg)
```

```nim
proc baz(c: int) =
  echo c
proc bar(b: int) =
  baz(b)
func fn1(a: int) =
  bar(a)
proc main()=
  fn1(3)
main()
```
current:
```
tests/nim/all/t12516.nim(9, 1) template/generic instantiation from here
tests/nim/all/t12516.nim(5, 6) Warning: 'baz' has side effects as it calls 'echo' [SideEffect]
  proc baz(c: int) =
       ^
tests/nim/all/t12516.nim(9, 1) template/generic instantiation from here
tests/nim/all/t12516.nim(7, 6) Warning: 'bar' has side effects as it calls 'baz' [SideEffect]
  proc bar(b: int) =
       ^
tests/nim/all/t12516.nim(9, 6) Error: 'fn1' has side effects as it calls 'bar'
  func fn1(a: int) =
```

instead something like this is better:
```
tests/nim/all/t12516.nim(9, 1) template/generic instantiation from here
tests/nim/all/t12516.nim(9, 6) Error: 'fn1' has side effects as it calls 'bar'
  tests/nim/all/t12516.nim(5, 6) 'baz' has side effects as it calls 'echo':  proc baz(c: int) =
  tests/nim/all/t12516.nim(7, 6) 'bar' has side effects as it calls 'baz': proc bar(b: int) =

  func fn1(a: int) =
```

Reply:
I changed `listSideEffects` to compile a single message.
I'm unsure of the actual rendered messages though.
The rendering form that you state above shows all locations as errors when I use the nim vscode plugin. But I think it would be clearer if the other locations were hints.

What about something like:
```
<lineinfo> Error: `symbol` has side effects
  <lineinfo> Hint: `symbol` accesses/mutates/calls `reason`
    <lineinfo> Hint: `reason` declared here
```

Reply:
one problem is the other locations are not really hints, they're mascarading as such; hints go through a specific code path in the compiler, which generates color, a hint name etc.

not a big deal though, we can always improve this later

Reply:
Yeah, realized that it's not the same, but I wanted the Nim VSCode plugin to be able to parse it and then markup my code in the editor. VSCode has diagnostic message types of hint, warning and error. The Nim VSCode plugin can be changed to map something else to internal VSCode hints.